### ISSUE TITLE
Pay method for testing environment

### ISSUE THREAD

**Issue by ghost** [w3c/browser-payment-api#613]
A "Pay" method is necessary for the UI testing environment to trigger the pay button.

**Comment by marcoscaceres** [comment]
Not sure I follow? `.show()` triggers showing the UI?

**Comment by ghost** [comment]
https://developers.google.com/web/fundamentals/discovery-and-monetization/payment-request/images/5_9_payment_request_ui.png

The blue pay button.

**Comment by marcoscaceres** [comment]
Unfortunately, we can't add that, as the web platform doesn't define a "UI testing environment" (i.e., in the same way we define a "SecureContext").  

As the person writing most of the tests for the test suite, I totally feel your pain tho. I'm having to test everything manually.

**Comment by marcoscaceres** [comment]
Note that we've tried to work around this - but we couldn't get consensus from everyone so we stopped :(:
https://marcoscaceres.github.io/payment-method-web-platform-test/

**Comment by ghost** [comment]
Ok, thanks a lot.

**Comment by marcoscaceres** [comment]
No problem. If you are testing this yourself, it be great to hear how you are working around this limitation (and degree of pain). 

If it becomes a issue for you, we'd like to hear it so we can try to address it in the future!

**Comment by ghost** [comment]
It will be appreciated if you can add the **test_pay** method. Then the payment gateway can follow your standard.

**Comment by rsolomakhin** [comment]
>  we couldn't get consensus from everyone

To be fair, there was no push back either, so I still plan to go down this path.

**Comment by ghost** [comment]
**test_pay** method --> environment':'TEST'

**Comment by marcoscaceres** [comment]
@sabrinatseng I just discovered that too by accident: 
https://docs.microsoft.com/en-us/microsoft-edge/dev-guide/device/payment-request-api

@rsolomakhin ^^^ "Test Mode in Microsoft Edge"... 

```JS 
var supportedInstruments = [{
    supportedMethods: ['basic-card'],
    data: {
        supportedNetworks: ['visa', 'mastercard', 'amex'],
        supportedTypes: ['credit'] 
        //for test transactions uncomment the test flag below
        //environment: 'TEST'
    }         
}]; 
```

It doesn't exactly allow us to automate things, but it's something.

**Comment by rsolomakhin** [comment]
Better than nothing. I'm OK implementing this. OK with you, @marcoscaceres?

**Comment by marcoscaceres** [comment]
>  OK with you, @marcoscaceres?

Yep, but I need to check if Edge has those addresses/values and test credit card details built-in. We should also get an assurance from Microsoft that they won't change it if we copy them.

If the above holds, we can easily define `environment: "TEST"` as a note in Basic Card.

**Comment by marcoscaceres** [comment]
@patrickkettner, @molly26dalton - see above. Would we be safe to copy Edge/Microsoft Wallet here?

**Comment by adrianhopebailie** [comment]
It would be nice to still have a test payment method so that this is not tied to `basic-card`.

**Comment by adrianhopebailie** [comment]
<meta>Lots of discussion here so re-opening</meta>

**Comment by marcoscaceres** [comment]
@adrianhopebailie if it's ok, I'd like us to move this to Basic Card, as it doesn't belong in Payment Request. I've opened: https://github.com/w3c/payment-method-basic-card/issues/41

**Comment by marcoscaceres** [comment]
@adrianhopebailie: 
> It would be nice to still have a test payment method so that this is not tied to basic-card.

Ah, sorry. I missed this comment. 

Yeah, it would be, but let's take baby steps... we [tried once](https://marcoscaceres.github.io/payment-method-web-platform-test/) already, but it proved too hard. So, the `environment` thing is a good start and simple enough to implement.

If we can get `environment` as a first step, then we can then return here and take the next step ("use TEST, and don't show the sheet... and behave like this...")

**Comment by ghost** [comment]
show method -> **shown event** -> **test_pay method** -> paymentResponse

**Comment by marcoscaceres** [comment]
@sabrinatseng, can you please write out what you mean more fully (or in JS)? I'm not able to follow your logic above.

**Comment by ghost** [comment]
Now the payment flow is **show method** -> **paymentResponse**.
```
var paymentRequest = new PaymentRequest( methodData, details, options );
paymentRequest.show().then(function(paymentResponse) {
        paymentResponse.complete('success');
})
```

It should add the paymentRequest shown event so the test_pay method can apply.
Something like this

```
paymentRequest.addEventListener('shown', function(){
        paymentRequest.test_pay()
})
```

**Comment by marcoscaceres** [comment]
Right, that won't work unfortunately (and we would need to add additional method and event). The model Microsoft has implemented gives us everything we need however (but with the limitation of needing to actually click on things, but that's ok for now). 

To overcome the "clicking on things" problem, we need a whole new payment method identifier (and model for driving it...).

**Comment by adrianhopebailie** [comment]
> Yeah, it would be, but let's take baby steps... we tried once already, but it proved too hard. So, the environment thing is a good start and simple enough to implement.

Let's mark this as a v2 feature and leave it open then. We're in the process of preparing a new charter and these postponed issues are our starting point.

**Comment by ghost** [comment]
Thank you @adrianhopebailie.

**Comment by marcoscaceres** [comment]
Can live with that. The problem space is effectively to control the main events and echo back requested data on those events and in the Payment response... and doing that in a secure way (e.g., only when dev tools is open or a browser pref is set, for automation).

**Comment by marcoscaceres** [comment]
Note that I'm using the label Future Candidate Feature also. I want us to move to an incremental release model using semantic release  (i.e., 1.1, 1.2, etc.). That way, we can spec, test, and land features more quickly, without getting bogged down with a monolithic "v2" release.

**Comment by ianbjacobs** [comment]
Hi @marcoscaceres 
That sounds like a great topic for TPAC. Can we carve out some time for discussion on versioning then?

Ian

**Comment by marcoscaceres** [comment]
Absolutely. We should have a fairly good idea of what we need by then. But track w3c/payment-method-basic-card#41, as it will serve as the basis for this (🤞).

**Comment by marcoscaceres** [comment]
Oh, you meant the 1.x thing... that too! ☺️

**Comment by ianbjacobs** [comment]
Yes 1.x. Added as note for agenda-building:
 https://github.com/w3c/webpayments/wiki/FTF-Nov2017
